### PR TITLE
8300493: Use ArraysSupport.vectorizedHashCode in j.u.zip.ZipCoder

### PR DESCRIPTION
**Pull request by cl4es**
`ZipCoder::checkedHashCode` emulates `StringLatin1::hashCode` but operates on a `byte[]` subrange. It can profitably use the recently introduced `ArraysSupport::vectorizedHashCode` method to see a speed-up, which translates to a small but significant speed-up on `ZipFile` creation.

Before:
```
Benchmark                     (size)  Mode  Cnt       Score      Error  Units
ZipFileOpen.openCloseZipFile     512  avgt   15   83007.325 ± 1446.716  ns/op
ZipFileOpen.openCloseZipFile    1024  avgt   15  154550.631 ± 2166.673  ns/op
```
After:
```
Benchmark                     (size)  Mode  Cnt       Score      Error  Units
ZipFileOpen.openCloseZipFile     512  avgt   15   79512.902 ±  814.449  ns/op
ZipFileOpen.openCloseZipFile    1024  avgt   15  147892.522 ± 2744.017  ns/op
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300493](https://bugs.openjdk.org/browse/JDK-8300493): Use ArraysSupport.vectorizedHashCode in j.u.zip.ZipCoder

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [0c7f0f4f](https://git.openjdk.org/jdk/pull/12077/files/0c7f0f4f700b8d3a01c853f065c24959caf75115)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12077/head:pull/12077` \
`$ git checkout pull/12077`

Update a local copy of the PR: \
`$ git checkout pull/12077` \
`$ git pull https://git.openjdk.org/jdk pull/12077/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12077`

View PR using the GUI difftool: \
`$ git pr show -t 12077`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12077.diff">https://git.openjdk.org/jdk/pull/12077.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
Using countPositives, and vectorizedHashCode(T_BOOLEAN) for unsigned bytes, make sense here. I don't have time to study the micro right now.

**Comment by AlanBateman** [comment]
> FWIW the micro is derived from the sibling `ZipFileGetEntry` micro in the same directory. It's not exactly necessary for this use case to add such a benchmark, but I think there's value in verifying that optimizing `checkedHash` improves `ZipFile` setup and adding the micro might allow us to find further opportunities down the line.

I've no doubt it improves checkedHash, it's just that open the zip file and reading in the CEN probably dominates when not in the file system cache.

**Review by AlanBateman — APPROVED** [review]

File: test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        try (FileOutputStream fos = new FileOutputStream(tempFile);
+             ZipOutputStream zos = new ZipOutputStream(fos)) {
+
+            Random random = new Random(4711);
+            for (int i = 0; i < size; i++) {
+                String ename = "long-directory-name-" + (random.nextInt(90000) + 10000) + "-" + i + "/";
+                zos.putNextEntry(new ZipEntry(ename));
+
+                ename += "long-entry-name-"  + (random.nextInt(90000) + 10000)  + "-" + i;
+                zos.putNextEntry(new ZipEntry(ename));
```
Comment (LanceAndersen) [inline test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java:69]: Would it be worth to add some random sized data when the entries are created to allow for getting a bit more insight, or  perhaps do that in a separate benchmark>

**Comment by cl4es** [comment]
> > FWIW the micro is derived from the sibling `ZipFileGetEntry` micro in the same directory. It's not exactly necessary for this use case to add such a benchmark, but I think there's value in verifying that optimizing `checkedHash` improves `ZipFile` setup and adding the micro might allow us to find further opportunities down the line.
> 
> I've no doubt it improves checkedHash, it's just that open the zip file and reading in the CEN probably dominates when not in the file system cache.

Right, the micro is a poor proxy for real-world implications since time to open a zip file very much depends on the filesystem speed but this is sort of by design. We have separate startup tests that tries to emulate more "cold start" scenarios, which micros like this are complementary to and not a substitute for.

File: test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java (hunk 1 above)
Comment (cl4es) [reply test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java:69]: I was experimenting with varying the entry names length to see what - if any - impact it had and saw a small effect on the micro. It does make more sense to vary lengths now that very long names will take different paths in the vectorized intrinsic. I'll see what I can do without overengineering this.

**Comment by cl4es** [comment]
Updated micro to vary entry sizes more to ensure we exercise the different code paths through the `hashCode` intrinsic. The new setup generates both longer and shorter entries than before, weighting up the average length a bit by increasing the spread. The longer entries see a proportionately larger speed-up, as expected since they benefit from vectorization. Removed some pointless randomness.

Baseline:
```
Benchmark                     (size)  Mode  Cnt       Score      Error  Units
ZipFileOpen.openCloseZipFile     512  avgt   15   98832.801 ± 2155.928  ns/op
ZipFileOpen.openCloseZipFile    1024  avgt   15  187373.545 ± 2296.779  ns/op
```

Patched:
```
Benchmark                     (size)  Mode  Cnt       Score      Error  Units
ZipFileOpen.openCloseZipFile     512  avgt   15   85574.648 ±  920.477  ns/op
ZipFileOpen.openCloseZipFile    1024  avgt   15  160493.277 ± 3450.928  ns/op
```

**Review by LanceAndersen — APPROVED** [review]
Thank you Claes

**Comment by cl4es** [comment]
Thanks @LanceAndersen and @AlanBateman for reviewing!

/integrate

**Pull request closed without merging**